### PR TITLE
e-books: holdings affiliated to libraries

### LINKED ISSUE #2902: Cantook e-books are affiliated to libraries
### User story
As a patron, I want to find the e-books of my library, so that I can download them on Cantook.

### Acceptance criteria
For RERO Valais

* for each e-book "mv-cantook", a holding or item is created for a new library called "En ligne / Online"
* the URL http://mediatheque-valais.cantookstation.eu stays as is

For RBNJ

* for each e-book "ebibliomedia", a holding or item is created for the libraries below
* it is possible to use the library facet to filter the e-books
* the URL https://bm.ebibliomedia.ch/… becomes according to the library:

BVCF - Bibl. de la Ville de La Chaux-de-Fonds > https://la-chaux-de-fonds.ebibliomedia.ch/…
BVL - Bibl. de la Ville du Locle > https://bv-lelocle.ebibliomedia.ch/…
BPUN Lecture publique > https://bpu-neuchatel.ebibliomedia.ch/…
Lycée cantonal de Porrentruy > https://lamedia-porrentruy.ebibliomedia.ch/…
BiCJ - Bibl. cantonale jurassienne > https://bicj-porrentruy.ebibliomedia.ch/…
BM Delémont > https://biblio-delemont.ebibliomedia.ch/…
CIP - Tramelan > https://cip-tramelan.ebibliomedia.ch/…

On a technical level, this implies

1. to have a harvesting config on library level (use case RBNJ), with cascade on organisation level (use case RERO Valais)
2. to have a harvesting config allowing multiple sources (for mid-term needs)

### PR DESCRIPTION
**Pull request by rerowep**
* Closes #2902

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

## Why are you opening this PR?

- Which task/US does it implement?
- Which issue does it fix?

## Dependencies

My PR depends on the following `rero-ils-ui`'s PR(s):

* rero/rero-ils-ui#<xx>

## How to test?

- What command should I have to run to test your PR?
- What should I test through the UI?

### CODE REVIEW

**Comment by pronguen** [comment]
- [ ] Why is there a difference between the two libraries?
    image

This is a great improvement! In a second step, this has to be tested in bib.test.rero.ch with the right configuration on real data.

File: rero_ils/modules/libraries/api.py [hunk 1]
```diff
@@ -385,6 +382,12 @@ def get_timezone(self):
         # TODO: Use BABEL_DEFAULT_TIMEZONE by default
         return pytz.timezone('Europe/Zurich')
 
+    def get_online_harvested_source_url(self, source):
+        """Get online harvested source url."""
```
Comment (zannkukai) [inline rero_ils/modules/libraries/api.py:386]: docstring about `source` param.

**Review by zannkukai — APPROVED** [review]

**Review by benerken — APPROVED** [review]

**Review by BadrAly — APPROVED** [review]

File: rero_ils/modules/ebooks/utils.py [hunk 2]
```diff
@@ -120,12 +126,18 @@ def update_document_holding(record, pid):
...
     for harvested_source in harvested_sources:
-        org = Organisation.get_record_by_online_harvested_source(
-            source=harvested_source['source'])
-        if org:
+        if org := Organisation.get_record_by_online_harvested_source(
+                source=harvested_source['source']):
             item_type_pid = org.online_circulation_category()
-            locations = org.get_online_locations()
-            for location_pid in locations:
+            location_pids = org.get_online_locations()
```
Comment (jma) [inline rero_ils/modules/ebooks/utils.py:132]: This code seems similar than the previous one. Can you create a method for this?

File: rero_ils/modules/libraries/jsonschemas/libraries/library-v0.0.1.json [hunk 3]
```diff
@@ -479,6 +479,36 @@
...
+      "type": "array",
+      "minItems": 1,
+      "items": {
+        "type": "object",
+        "additionalProperties": false,
+        "required": [
+          "source",
+          "url"
+        ],
+        "properties": {
```
Comment (jma) [inline rero_ils/modules/libraries/jsonschemas/libraries/library-v0.0.1.json:495]: OrderedProperties is missing.

File: tests/api/organisations/test_organisations_rest_api.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ def test_get_record_by_viewcode(org_martigny):
 
 def test_get_record_by_online_harvested_source(org_martigny):
     """Test get_record_by_online_harvested_source."""
-    source = org_martigny.get('online_harvested_source')
+    source = org_martigny.get('online_harvested_source')[0]
```
Comment (jma) [inline tests/api/organisations/test_organisations_rest_api.py:40]: You can probably use pop instead?

**Review by jma — APPROVED** [review]

**Pull request merged**